src/snowflake/cli/api/utils/rendering.py

```diff
@@ -88,24 +97,138 @@ def jinja_render_from_file(
         print(rendered_result)
 
 
-class _AttrGetter:
-    def __init__(self, data_dict):
-        self._data_dict = data_dict
+def _add_project_context(
+    external_data: Dict, project_definition: ProjectDefinition
+) -> Dict:
+    """
+    Updates the external data with variables from snowflake.yml definition file.
+    """
+    if _CONTEXT_KEY in external_data:
+        raise ClickException(
+            f"{_CONTEXT_KEY} in user defined data. The `{_CONTEXT_KEY}` variable is reserved for CLI usage."
+        )
+
+    context_data = _resolve_variables_in_project(project_definition)
+
+    context_data.update(external_data)
+    return context_data
 
-    def __getattr__(self, item):
-        if item not in self._data_dict:
-            raise AttributeError(f"No attribute {item}")
-        return self._data_dict[item]
 
+def _remove_ctx_env_prefix(text: str) -> str:
+    prefix = "ctx.env."
+    if text.startswith(prefix):
+        return text[len(prefix) :]
+    return text
 
-def _add_project_context(data: Dict):
-    context_key = "ctx"
-    if context_key in data:
-        raise ValueError(f"{context_key} in user defined data")
-    context_data = {context_key: {"env": _AttrGetter(os.environ)}}
-    return {**data, **context_data}
+
+def string_includes_template(text: str) -> bool:
+    return bool(re.search("&{.+}", text))
+
+
+def _resolve_variables_in_project(project_definition: ProjectDefinition):
+    # If there's project definition file then resolve variables from it
+    if not project_definition or not project_definition.meets_version_requirement(
+        "1.1"
+    ):
+        return {_CONTEXT_KEY: {"env": DictWithEnvironFallback({})}}
+
+    variables_data: DictWithEnvironFallback = cast(
+        DictWithEnvironFallback, project_definition.env
+    )
+
+    unresolved_keys: List[str] = _check_variables_consistency(variables_data)
+    env = get_snowflake_cli_jinja_env()
+    context_data = {_CONTEXT_KEY: project_definition}
+    while unresolved_keys:
+        key = unresolved_keys.pop()
+        value = variables_data[key]
+        if not isinstance(value, str):
+            continue
+        try:  # try to evaluate the template given current state of know variables
+            variables_data[key] = env.from_string(value).render(context_data)
+            if string_includes_template(variables_data[key]):
+                unresolved_keys.append(key)
+        except UndefinedError:
+            unresolved_keys.append(key)
+
+    return context_data
+
+
+def _check_for_cycles(nodes: defaultdict):
+    nodes = nodes.copy()
+    for key in list(nodes):
+        q = deque([key])
+        visited: List[str] = []
+        while q:
+            curr = q.popleft()
+            if visited and curr == key:
```

Comment:
you probably need to check if curr in visited instead. What if you have a -> b -> c -> b -> c -> b ... this might cause infinite loop (I think)

Reply:
good point, fixed